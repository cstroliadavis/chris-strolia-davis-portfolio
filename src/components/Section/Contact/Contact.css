.section.contact {
    background-color: #98c4c5;
}

.section.contact > .container {
    margin: auto;
    width: 400px;
    text-align: left;
}

.section.contact .name,
.section.contact .email,
.section.contact .message {
    margin: 1rem;
    width: 100%;
}

.section.contact input,
.section.contact textarea {
    width: 100%;
}

.section.contact .error {
    color: red;
    display: none;
}

.section.contact .error.show {
    display: block;
}

.section.contact .message-sent {
    display: none;
    text-align: center;
    color: darkgreen;
}

.section.contact .message-sent.show {
    display: block;
}

.section.contact #contact-form button[type="submit"] {

}
