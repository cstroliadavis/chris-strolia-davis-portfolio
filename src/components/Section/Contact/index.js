import './Contact.css';
import { useState } from "react";
import { validateEmail } from '../../../utils/string-helpers';
import { contactUs } from "../../../services/contact";

const pause = (millis) => new Promise((resolve) => setTimeout(resolve, millis));

const defaultFormFields = { name: '', email: '', message: '' };
const defaultErr = {
  showNameRequired: '',
  showEmailRequired: '',
  showEmailMustBeValid: '',
  showMessageRequired: '',
};

export default function Contact() {
  const [ formState, setFormState ] = useState(defaultFormFields)
  const [ messageSent, setMessageSent ] = useState('');
  const [ err, setErr ] = useState(defaultErr);
  const [ submitDisabled, setSubmitDisabled ] = useState(true);
  const { name, email, message } = formState;

  function formNotReady({ name, email, message }) {
    return !(name && email && validateEmail(email) && message);
  }

  function cleanErrors({ name, email, message }) {
    const cleared = {}
    if (name) cleared.showNameRequired = '';
    if (email) cleared.showEmailRequired = '';
    if (validateEmail(email)) cleared.showEmailMustBeValid = '';
    if (message) cleared.showMessageRequired = '';

    setErr({ ...err, ...cleared });
  }

  const validations = {
    email(value) {
      setErr({
        ...err,
        showEmailRequired: value ? '':'show',
        showEmailMustBeValid: value && !validateEmail(value) ? 'show':'',
      });
    },
    name(value) {
      setErr({ ...err, showNameRequired: value ? '':'show' });
    },
    message(value) {
      setErr({ ...err, showMessageRequired: value ? '':'show' });
    },
  };

  function modifyInput(e) {
    const { name, value } = e.target;
    const newVals = { ...formState, [name]: value };
    cleanErrors(newVals);
    setFormState(newVals);
    setSubmitDisabled(formNotReady(newVals));
  }

  const inputEvents = {
    onBlur(e) {
      const { name: target, value } = e.target;

      const action = validations[target];
      action(value);

      setSubmitDisabled(formNotReady({ name, email, message }))
    },
    onChange: modifyInput,
    onInput: modifyInput,
  }

  function handleSubmit(e) {
    e.preventDefault();
    contactUs(formState)
      .then(() => setMessageSent('show'))
      .then(() => pause(5000))
      .then(() => setMessageSent(''));
  }

  return (
    <div className="container">
      <h1>Contact me</h1>
      <form id="contact-form" onSubmit={ handleSubmit }>
        <div className="name">
          <input type="text" name="name" placeholder="Name" defaultValue={ name } { ...inputEvents }/>
          <div className={ `error ${ err.showNameRequired }` }>Please enter a name.</div>
        </div>
        <div className="email">
          <input type="email" name="email" placeholder="E-mail" defaultValue={ email } { ...inputEvents }/>
          <div className={ `error ${ err.showEmailRequired }` }>Please enter an email.</div>
          <div className={ `error ${ err.showEmailMustBeValid }` }>Please enter a valid email address.</div>
        </div>
        <div className="message">
          <textarea name="message" placeholder="Message" rows="5" defaultValue={ message } { ...inputEvents }/>
          <div className={ `error ${ err.showMessageRequired }` }>Please include a message.</div>
        </div>
        <div className={ `message-sent ${ messageSent }` }>Message was sent successfully.</div>
        <button disabled={ submitDisabled } type="submit">Submit</button>
      </form>
    </div>
  )
}
