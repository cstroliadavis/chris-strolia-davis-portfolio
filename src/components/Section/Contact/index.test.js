import { render, cleanup } from '@testing-library/react';
import Contact from './index'

afterEach(cleanup);

describe('Contact', () => {
  it('renders', () => {
    render(<Contact/>)
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<Contact/>);

    expect(asFragment()).toMatchSnapshot();
  });

});
