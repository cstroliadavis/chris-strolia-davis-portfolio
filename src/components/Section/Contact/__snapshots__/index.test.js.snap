// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1>
      Contact me
    </h1>
    <form
      id="contact-form"
    >
      <div
        class="name"
      >
        <input
          name="name"
          placeholder="Name"
          type="text"
          value=""
        />
        <div
          class="error "
        >
          Please enter a name.
        </div>
      </div>
      <div
        class="email"
      >
        <input
          name="email"
          placeholder="E-mail"
          type="email"
          value=""
        />
        <div
          class="error "
        >
          Please enter an email.
        </div>
        <div
          class="error "
        >
          Please enter a valid email address.
        </div>
      </div>
      <div
        class="message"
      >
        <textarea
          name="message"
          placeholder="Message"
          rows="5"
        />
        <div
          class="error "
        >
          Please include a message.
        </div>
      </div>
      <div
        class="message-sent "
      >
        Message was sent successfully.
      </div>
      <button
        disabled=""
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
